### PR TITLE
Generate worker fuse helm yaml files

### PR DESCRIPTION
**Pull request by ssz1997**
This PR is a continuation of [https://github.com/Alluxio/alluxio/pull/13234](url)

To enable FUSE in worker container in k8s environment, the helm yaml files need to be modified. According to [https://github.com/Alluxio/alluxio/pull/13234](url), instead of directly providing ready-to-use templates, we want to auto-generate these templates with helm-generate.sh. User needs to include "--enable-worker-fuse" in command line to turn worker fuse on. 
The first commit is able to generate workerFuse container yaml files if .Values.worker.fuseEnabled is true.
The second commit is able to get the parameter whether to enable workerFuse from command line and pass it. 

Fix Alluxio/new-contributor-tasks#587

### CODE REVIEW

File: integration/kubernetes/helm-chart/alluxio/templates/config/alluxio-conf.yaml
Comment (LuQQiu) [inline integration/kubernetes/helm-chart/alluxio/templates/config/alluxio-conf.yaml:200]: this is the end of file line? every file should have an empty line at the end of the file

File: integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml
Comment (LuQQiu) [inline integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml:252]: Using
```
        - name: alluxio-fuse-mount
          hostPath:
            path: {{ .Values.fuse.mountPath | dir }}
            type: DirectoryOrCreate
```
similar to fuse/daemonset.yaml?

File: integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml
Comment (LuQQiu) [inline integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml:159]: Using
```
            - name: alluxio-fuse-mount
              mountPath: {{ .Values.fuse.mountPath | dir }}
              mountPropagation: Bidirectional
```
similar to fuse/daemonset.yaml?

File: integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml
Comment (LuQQiu) [inline integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml:55]: just curious, what's the differences with and without `-`

File: integration/kubernetes/helm-chart/alluxio/templates/config/alluxio-conf.yaml
Comment (LuQQiu) [inline integration/kubernetes/helm-chart/alluxio/templates/config/alluxio-conf.yaml:88]: Use .Values.fuse.mountOptions to replace a hard coded `allow_other` value?

**Review by LuQQiu — COMMENTED** [review]
Thanks for supporting worker fuse in helm files! @jiacheliu3 PTAL, thanks

File: integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml
Comment (jiacheliu3) [reply integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml:55]: The `-`  at start of line will delete the whitespace on the previous line. `-` at the end of line will delete the following whitespace on the same line.
See
https://helm.sh/docs/chart_template_guide/control_structures/#controlling-whitespace

Comment (ssz1997) [reply integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml:55]: Without the two - here, the generated yaml template will have an empty line there

File: integration/kubernetes/helm-chart/alluxio/templates/config/alluxio-conf.yaml
Comment (LuQQiu) [inline integration/kubernetes/helm-chart/alluxio/templates/config/alluxio-conf.yaml:88]: I do not quite understand why we use .Values.fuse.mountOptions like that

Comment (LuQQiu) [reply integration/kubernetes/helm-chart/alluxio/templates/config/alluxio-conf.yaml:88]: I would prefer to use one line for all fuse mount options like allow_other,kernel_cache,attr_timeout=9000.
and use
```
{{- $workerJavaOpts = print "-Dalluxio.worker.fuse.mount.options=.Values.fuse.mountOptions" | append $workerJavaOpts }}
```
here

File: integration/kubernetes/helm-chart/alluxio/values.yaml
Comment (LuQQiu) [inline integration/kubernetes/helm-chart/alluxio/values.yaml:314]: fuse options specific to fuse embedded in the worker process?

**Review by LuQQiu — COMMENTED** [review]
Left some minor comments, otherwise LGTM.

@ssz1997 A small tip for future PRs is to reply to the comments. This PR is fine, there are not that many comments. For a PR with tons of comments, it will be easier for the reviewer to know what changes you make. Or reply to the most important comments or the comments need discussion

File: integration/kubernetes/helm-chart/alluxio/templates/config/alluxio-conf.yaml
Comment (ssz1997) [reply integration/kubernetes/helm-chart/alluxio/templates/config/alluxio-conf.yaml]: I modified it to be how jvmOptions work. The options would look clearer as an array on multiple lines in values.yaml, just like jvmOptions. Then using concat instead of loop and append can make them all on one line. Do you think this should be the way to go?

File: integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml
Comment (jiacheliu3) [reply integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml:55]: We used to carefully get rid of all the extra empty lines. But later we found it's really hard to guarantee there's no extra empty lines in the generated YAMLs so we implicitly loosed the standard to "definitely avoid empty whitespace that breaks the YAML indentation and try your best to avoid unnecessary empty line".

File: integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml:210]: Good work removing extra whitespace

File: integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml:253]: Note that `hostPath` that are created by `DirectoryOrCreate` are only writable by `root` user
Ref: https://kubernetes.io/docs/concepts/storage/volumes/#hostpath

File: integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml:114]: I'm not sure if this should be the hostIP or the podIP, pls double check if this works, especially if this will enable the process-local read on the worker pod @LuQQiu

File: integration/kubernetes/helm-chart/alluxio/templates/config/alluxio-conf.yaml
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/templates/config/alluxio-conf.yaml:169]: Would `{{- if eq .Values.worker.fuseEnabled false }}` be better?

File: integration/kubernetes/helm-chart/alluxio/templates/config/alluxio-conf.yaml
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/templates/config/alluxio-conf.yaml:207]: Same comment as above

File: integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml
Comment (LuQQiu) [reply integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml:114]: emmm, it should be similar to the ALLUXIO_WORKER_HOSTNAME, 
```
          env:
          - name: ALLUXIO_WORKER_HOSTNAME
            valueFrom:
              fieldRef:
                fieldPath: status.hostIP
          {{- if not .Values.worker.hostNetwork }}
          - name: ALLUXIO_WORKER_CONTAINER_HOSTNAME
            valueFrom:
              fieldRef:
                fieldPath: status.podIP
```
i saw we use hostIP for worker hostname. This works in my testing.
Not quite sure why we need a new property key `ALLUXIO_WORKER_CONTAINER_HOSTNAME` in container env?
should i create a similar `ALLUXIO_CLIENT_CONTAINER_HOSTNAME`?

File: integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml
Comment (ssz1997) [reply integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml:253]: @LuQQiu

Comment (LuQQiu) [reply integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml:253]: Emmm... by using fuse, the pod needs to run as root
```
      securityContext:
        runAsUser: 0
        runAsGroup: 0
        fsGroup: 0
```
so that it's fine that `DirectoryOrCreate` can only writable by root user.
And interestingly, the mount point is actually accessible by other users in my testing. 
I can double-check this part

**Comment by ssz1997** [comment]
@LuQQiu PTAL. Thanks

File: integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml
Comment (LuQQiu) [reply integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml:264]: We can change to Directory here. in our testing, we create the mount folder first

File: integration/kubernetes/helm-generate.sh
Comment (LuQQiu) [inline integration/kubernetes/helm-generate.sh:27]: OPTS the options for generating desired YAML template. Currently...

File: integration/kubernetes/helm-generate.sh
Comment (LuQQiu) [inline integration/kubernetes/helm-generate.sh:28]: --worker-fuse Launch FUSE application in the worker process inside worker container.

File: integration/kubernetes/helm-chart/alluxio/templates/config/alluxio-conf.yaml
Comment (ssz1997) [reply integration/kubernetes/helm-chart/alluxio/templates/config/alluxio-conf.yaml:200]: Isn't this because gitHub doesn't show the empty line at the end? I can see the empty line in ItelliJ.

File: integration/kubernetes/helm-generate.sh
Comment (ssz1997) [inline integration/kubernetes/helm-generate.sh:202]: @LuQQiu Since we only have one option, so directly compare $3 to "--worker-fuse" is fine. But if we have more options, we should not do this but use an array storing all options and use switch statement inside the loop. Do you think we should do this?

**Comment by ssz1997** [comment]
@LuQQiu PTAL. Thanks!

File: integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml
Comment (LuQQiu) [reply integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml:114]: @ssz1997 If you search `alluxio.worker.container.hostname` in Alluxio property key file, you will find that property key, but there is no such a thing as `alluxio.client.container.hostname`.
Next time if you make any change, make sure you understand why the changes are needed
e.g. why we use `ALLUXIO_WORKER_CONTAINER_HSOTNAME` `alluxio.worker.container.hostname` in alluxio code base. and then think if we should add a similar `alluxio.client.container.hostname` in the alluxio java code.
Adding `ALLUXIO_CLIENT_CONTAINER_HOSTNAME` here doesn't make any sense since no application consumes this information.

File: integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml
Comment (jiacheliu3) [reply integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml:81]: +1 to @ZhuTopher , let's err out when a discrepancy is spotted, instead of letting the container run into permission issues at runtime.

Also don't forget to add to the comments, as previously mentioned :)

**Comment by ssz1997** [comment]
@jiacheliu3 @ZhuTopher I added a worker pod security context, and users should modify them to 0 if they want to embed Fuse in worker process. I feel like this makes more sense because in this way the worker pod configuration is getting the values from .Values.worker instead of .Values.fuse, which is used for standalone Fuse pod. PTAL, thanks!

File: integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml
Comment (ZhuTopher) [inline integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml]: Doesn't this implementation prevent the worker pods from using the deployment-wide defaults in `.Values.(user|group|fsGroup)`?

I'd prefer for `.Values.worker.(user|group|fsGroup)` to override `.Values.(user|group|fsGroup)`, and have the default in `values.yaml` be empty or some other way to indicate to use the global values. Eg:

```
worker:
  user:    # supply a value to override the top-level defaults
  group:   # supply a value to override the top-level defaults
  fsGroup: # supply a value to override the top-level defaults
```

```
securityContext:
  {{- if .Values.worker.user }}
  runAsUser: {{ .Values.worker.user }}
  {{- else }}
  runAsUser: {{ .Values.user }}
  {{- end }}
```

You'd also need to change your implementation of checking that the assigned securityContext is running with UID/GID `0` too.

File: integration/kubernetes/helm-chart/alluxio/values.yaml
Comment (ZhuTopher) [inline integration/kubernetes/helm-chart/alluxio/values.yaml:238]: nit: Phrasing/wording

```
Setting fuseEnabled to true will embed FUSE in the worker process.
The worker pods will run the Alluxio worker using privileged containers,
with the `SYS_ADMIN` capability added.
Be sure to give root access to the pod by setting the worker Pod's
securityContext values to `0`.
```

**Review by ZhuTopher — CHANGES_REQUESTED** [review]

File: integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml
Comment (ssz1997) [reply integration/kubernetes/helm-chart/alluxio/templates/worker/daemonset.yaml]: Thanks for the clarification! I misunderstood what you meant yesterday.

**Comment by ssz1997** [comment]
@jiacheliu3 @ZhuTopher PTAL. Thanks!

**Review by ZhuTopher — APPROVED** [review]
Thanks for the changes, Shawn! Let's wait for Jiacheng's review as well before merging @LuQQiu

**Review by jiacheliu3 — APPROVED** [review]
LGTM, thx for the improvements!

**Comment by jiacheliu3** [comment]
The failing test is irrelevant to the change in the helm chart

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Pull request merged**
